syntax = "proto3";
package demo;


message Key {
    int64 key = 1;
}

message Value {
    int64 version = 1;
    int64 timestamp = 2;
    bytes data = 3;
}

message KeyValue {
    Key key = 1;
    Value value = 2;
  }

message RequestSET {
    Key key = 1;
    int64 timestamp = 2;
    bytes data = 3;
}

message ResponseSET {
    string status = 1;  // SUCCESS ou ERROR
    Value value = 2; // pode ser NULL
}

message RequestGET {
    Key key = 1;
}

message ResponseGET {
    string status = 1;  // SUCCESS ou ERROR
    Value value = 2; // pode ser NULL
}

message RequestDEL {
    Key key = 1;
    int64 version = 2; // pode ser NULL
}

message ResponseDEL {
    string status = 1;  // SUCCESS, ERROR, ERROR_NE ou ERROR_WV
    Value value = 2; // pode ser NULL
}

message RequestTESTANDSET {
    KeyValue keyValue = 1;
    int64 version = 2; // pode ser NULL
}

message ResponseTESTANDSET{
    string status = 1;  // SUCCESS, ERROR, ERROR_NE ou ERROR_WV
    Value value = 2; // pode ser NULL
}

service DBService {
    rpc SET (RequestSET) returns (ResponseSET) {}
    rpc GET (RequestGET) returns (ResponseGET) {}
    rpc DEL (RequestDEL) returns (ResponseDEL) {}
    rpc TESTANDSET (RequestTESTANDSET) returns (ResponseTESTANDSET) {}
  }


